src/providers/openstack/configdrive.rs

```diff
@@ -0,0 +1,209 @@
+            temp_dir: Some(target),
+        };
+        Ok(cd)
+    }
+
+    /// Return the path to the metadata directory.
+    fn metadata_dir(&self, platform: &str) -> PathBuf {
+        self.drive_path.clone().join(platform).join("latest")
+    }
+
+    /// Parse metadata attributes
+    ///
+    /// Metadata file contains a JSON object, corresponding to `MetadataEc2JSON`.
+    fn parse_metadata_ec2<T: Read>(input: BufReader<T>) -> Result<MetadataEc2JSON> {
+        serde_json::from_reader(input).chain_err(|| "failed parse JSON metadata")
+    }
+
+    /// Parse metadata attributes
+    ///
+    /// Metadata file contains a JSON object, corresponding to `MetadataOpenstackJSON`.
+    fn parse_metadata_openstack<T: Read>(input: BufReader<T>) -> Result<MetadataOpenstackJSON> {
+        serde_json::from_reader(input).chain_err(|| "failed parse JSON metadata")
+    }
+
+    /// The metadata is stored as key:value pair in ec2/latest/meta-data.json file
+    fn read_metadata(&self) -> Result<MetadataEc2JSON> {
+        let filename = self.metadata_dir("ec2").join("meta-data.json");
+        let file =
+            File::open(&filename).chain_err(|| format!("failed to open file '{:?}'", filename))?;
+        let bufrd = BufReader::new(file);
+        Self::parse_metadata_ec2(bufrd)
+            .chain_err(|| format!("failed to parse file '{:?}'", filename))
+    }
+
+    /// The public key is stored as key:value pair in openstack/latest/meta_data.json file
+    fn fetch_publickeys(&self) -> Result<Vec<PublicKey>> {
+        let filename = self.metadata_dir("openstack").join("meta_data.json");
+        let file =
+            File::open(&filename).chain_err(|| format!("failed to open file '{:?}'", filename))?;
+
+        let bufrd = BufReader::new(file);
+        let metadata: MetadataOpenstackJSON = Self::parse_metadata_openstack(bufrd)
+            .chain_err(|| format!("failed to parse file '{:?}'", filename))?;
+
+        let public_keys_map = metadata.public_keys.unwrap_or_default();
+        let public_keys_vec: Vec<&std::string::String> = public_keys_map.values().collect();
+        let mut out = vec![];
+        for key in public_keys_vec {
+            let key = PublicKey::parse(key)?;
+            out.push(key);
+        }
+        Ok(out)
+    }
+}
+
+impl MetadataProvider for OpenstackConfigDrive {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let metadata: MetadataEc2JSON = self.read_metadata()?;
+        Ok(maplit::hashmap! {
+            "OPENSTACK_HOSTNAME".to_string() => metadata.hostname.unwrap_or_default(),
+            "OPENSTACK_INSTANCE_ID".to_string() => metadata.instance_id.unwrap_or_default(),
+            "OPENSTACK_INSTANCE_TYPE".to_string() => metadata.instance_type.unwrap_or_default(),
+            "OPENSTACK_IPV4_LOCAL".to_string() => metadata.local_ipv4.unwrap_or_default(),
+            "OPENSTACK_IPV4_PUBLIC".to_string() => metadata.public_ipv4.unwrap_or_default(),
```

Comment:
I think we can skip all the `None` values instead of producing empty-string values.

Reply:
Sure, updated with the `is_some()` checking.

Reply:
What about using `if let Some(val) = metadata.foo { ... }` instead? That would avoid all the `unwrap`s.